### PR TITLE
Balder/use hash2d

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR
@bratseth FYI
Similar to #24935

### CODE REVIEW

File: eval/src/vespa/eval/eval/basic_nodes.h [hunk 1]
```diff
@@ -120,9 +121,9 @@ class String : public Leaf {
 public:
     String(const vespalib::string &value_in) : _value(value_in) {}
     bool is_const_double() const override { return true; }
-    double get_const_double_value() const override { return hash(); }
+    double get_const_double_value() const override { return hash2d(_value); }
     const vespalib::string &value() const { return _value; }
-    uint32_t hash() const { return hash_code(_value.data(), _value.size()); }
+    uint64_t hash() const { return hashValue(_value.data(), _value.size()); }
```
Comment (havardpe) [inline eval/src/vespa/eval/eval/basic_nodes.h:126]: this needs to agree with get_const_double_value or go away

File: vespalib/src/vespa/vespalib/util/string_hash.cpp [hunk 2]
```diff
@@ -1,19 +1,23 @@
...
-        hash = (hash << 5) - hash + str[i];
+double hash2d(const char *str, size_t len) {
+    size_t h = hashValue(str, len);
+    if ((h & 0x7ff0000000000000ul) == 0x7ff0000000000000ul) {
+        // Avoid nan
+        h = h & 0xffeffffffffffffful;
     }
-    return hash;
+    double d = 0;
+    memcpy(&d, &h, sizeof(d));
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/string_hash.cpp:15]: should have static_assert that size_t is the same size as a double

**Review by havardpe — COMMENTED** [review]
I am not sure if changing the hash function will break stuff or not. Hopefully not.

File: vespalib/src/vespa/vespalib/util/string_hash.cpp (hunk 2 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/string_hash.cpp:16]: Fixed

File: eval/src/vespa/eval/eval/basic_nodes.h (hunk 1 above)
Comment (baldersheim) [reply eval/src/vespa/eval/eval/basic_nodes.h]: Using get_const_double_value to generate the key and removed hash method.

**Pull request merged**